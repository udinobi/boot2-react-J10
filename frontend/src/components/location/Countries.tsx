import React from "react"

// import styled from "styled-components"

class Countries extends React.Component {

    public render() {
        return (
            <h1 className="title">Please, enter the location you want the weather report for</h1>
        )
    }
}

export default Countries
