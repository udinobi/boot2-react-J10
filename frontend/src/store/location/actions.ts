import { ActionCreator } from "redux"

import { Location } from "./types"

import { LocationChangedAction } from "./types"

export const changeCurrentLocation: ActionCreator<LocationChangedAction> = (location: Location) => ({
    payload: {
        location,
    },
    type: "location/LOCATION_CHANGED"
})
