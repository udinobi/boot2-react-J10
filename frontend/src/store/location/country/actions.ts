import axios, { AxiosResponse } from "axios"
import { ActionCreator, Dispatch } from "redux"

import {
    CountriesFailureAction, CountriesLoadingAction, CountriesRetrievedAction,
    Country, CountryActions, CountryActionType,
    CountrySelectedAction, CountryState
} from "./types"

// CountriesFailureAction --------------------------------------------------------------------------

const countriesFailure: ActionCreator<CountriesFailureAction> = () => ({
    type: CountryActionType.COUNTRIES_FAILURE
})

// CountriesLoadAction -----------------------------------------------------------------------------

const countriesLoading: ActionCreator<CountriesLoadingAction> = () => ({
    type: CountryActionType.COUNTRIES_LOADING
})

// Not checking yet, at this stage in our CSR app, if these properties are set or not.
// In a future SSR rendering feature branch we can also consider to implement logging
// to the server-side RT.
const url =
    `${process.env.REACT_APP_SERVICE_URL}${process.env.REACT_APP_LOAD_COUNTRIES_PATH}`

export const loadCountries = () =>
    async (dispatch: Dispatch<CountryActions>) => {
        dispatch(countriesLoading())

        try {
            const response: AxiosResponse<CountryState> = await axios.get(url)
            dispatch(countriesRetrieved(response.data))
        } catch (error) {
            dispatch(countriesFailure())
        }
    }
        
// CountriesRetrievedAction ------------------------------------------------------------------------

const countriesRetrieved: ActionCreator<CountriesRetrievedAction> = (countries: Country[]) => ({
    payload: { countries },
    type: CountryActionType.COUNTRIES_RETRIEVED,
})

// CountrySelectedAction ---------------------------------------------------------------------------

export const countrySelectedAction: ActionCreator<CountrySelectedAction> = (country: Country) => ({
    payload: { country },
    type: CountryActionType.COUNTRY_SELECTED
})

export const countrySelected = (country: Country) =>
    (dispatch: Dispatch<CountrySelectedAction>) => dispatch(countrySelectedAction(country))

// Action mapping ----------------------------------------------------------------------------------

export interface CountriesProps {
    countries: Country[]
    countrySelected: typeof countrySelected
    loadCountries: typeof loadCountries
}

export const mapCountryDispatchToProps = {
    countrySelected,
    loadCountries
}
    