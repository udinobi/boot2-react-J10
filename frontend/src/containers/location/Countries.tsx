import { Select } from "antd"
import React from "react"
import { connect } from "react-redux"

import styled from "styled-components"

import { None, Option } from "tsoption"

import { AppState } from "../../store/store"

import {
    LoadCountriesDispatchProps,
    mapCountriesDispatchToProps
} from "../../store/location/countriesActions"

import { CountriesState, Country } from "../../store/location/types"

interface SelectionState {
    country: Option<Country>
}

interface Value {
    key: string
    label: string
}

class Countries extends React.Component<LoadCountriesDispatchProps, CountriesState & SelectionState> {

    private readonly Option = Select.Option

    private readonly select = styled(Select)`
        font-size: 1.1rem;
        width: 100%;
    `

    constructor(props: LoadCountriesDispatchProps) {
        super(props)
        this.state = { countries: [], country: None.of() }
    }

    public componentDidMount() {
        this.props.loadCountries()
    }

    public componentWillReceiveProps(props: LoadCountriesDispatchProps & CountriesState) {
        if (this.state.countries !== props.countries) {
            const country = props.countries.length === 0
                ? Option.of<Country>()               // No countries... need to clear <select>
                : this.state.country.flatMap(_1 =>   // Have countries and if have current selection then... 
                    // check if selection is in props.countries
                    Option.of(props.countries.find(_2 => _2.code === _1.code))
                )
                // if not, set the <select>'s value to the 1st element in props.countries
                .orElse(Option.of(props.countries[0]))

            country.map(_ => props.countrySelected(_))

            this.setState({
                countries: props.countries,
                country
            })
        }
    }

    public render() {
        const countries = this.state.countries;
        const options = countries.map(country =>
            <this.Option value={country.code} key={country.name}>
                {country.name}
            </this.Option>
        );

        const value = this.state.country.map(country => ({
            key: country.code,
            label: country.name
        })).getOrElse({ key: "", label: "" })

        return (
            <this.select
                labelInValue={true}
                onSelect={this.onSelect}
                value={value}
            >
                {options}
            </this.select>
        )
    }

    private readonly onSelect = (value: Value) => {
        const country = {
            code: value.key,
            name: value.label
        }
        this.setState({ country: Option.of(country) })
        this.props.countrySelected(country)
    }
}

const mapStateToProps = (state: AppState): CountriesState => ({
    countries: state.countriesState.countries
})

export default connect(mapStateToProps, mapCountriesDispatchToProps)(Countries);
