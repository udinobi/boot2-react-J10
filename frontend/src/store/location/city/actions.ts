import { ActionCreator, Dispatch } from "redux"

import {
    CityActionType,
    Location, LocationSelectedAction,
    SuggestionsFailureAction, SuggestionsLookupAction,
    SuggestionsResetAction, SuggestionsRetrievedAction
} from "./types"

import { Country } from "../country/types"

// LocationSelectedAction --------------------------------------------------------------------------

const locationSelectedAction: ActionCreator<LocationSelectedAction> = (location: Location) => ({
    payload: { location },
    type: CityActionType.LOCATION_SELECTED
})

export const locationSelected = (location: Location) =>
    (dispatch: Dispatch<LocationSelectedAction | SuggestionsResetAction>) => {
        dispatch(suggestionsReset())
        dispatch(locationSelectedAction(location))
    }

// SuggestionsFailureAction ------------------------------------------------------------------------

export const suggestionsFailure: ActionCreator<SuggestionsFailureAction> = () => ({
    type: CityActionType.SUGGESTIONS_FAILURE
})

// SuggestionsLookupAction -------------------------------------------------------------------------

export const suggestionsLookup: ActionCreator<SuggestionsLookupAction> = (country: Country, locationTerm: string) => ({
    payload: { country, locationTerm },
    type: CityActionType.SUGGESTIONS_LOOKUP
})

// SuggestionsResetAction ------------------------------------------------------------------------

const suggestionsReset: ActionCreator<SuggestionsResetAction> = () => ({
    type: CityActionType.SUGGESTIONS_RESET
})

// SuggestionsRetrievedAction ----------------------------------------------------------------------

export const suggestionsRetrieved: ActionCreator<SuggestionsRetrievedAction> = (country: Country, suggestions: Location[]) => ({
    payload: { country, suggestions },
    type: CityActionType.SUGGESTIONS_RETRIEVED
})

// Action mapping ----------------------------------------------------------------------------------

export interface CityDispatchProps {
    locationSelected: typeof locationSelected
    suggestionsLookup: typeof suggestionsLookup
    suggestionsReset: typeof suggestionsReset
}

export const mapCityDispatchToProps = {
    locationSelected,
    suggestionsLookup,
    suggestionsReset
}
