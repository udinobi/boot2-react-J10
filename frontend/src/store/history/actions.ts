import { ActionCreator, Dispatch } from "redux"

import { Location } from "../location/city/types"

import {
    HistoryActionType, HistoryItem, LocationReloadAction, LocationRemoveAction
} from './types'

// LocationReloadAction ----------------------------------------------------------------------------

const locationReloadAction: ActionCreator<LocationReloadAction> = (history: HistoryItem) => ({
    payload: { history },
    type: HistoryActionType.LOCATION_RELOAD
})

export const reloadLocation = (history: HistoryItem) =>
    (dispatch: Dispatch<LocationReloadAction>) => dispatch(locationReloadAction(history))

// LocationRemoveAction ----------------------------------------------------------------------------

const locationRemoveAction: ActionCreator<LocationRemoveAction> = (location: Location) => ({
    payload: { location },
    type: HistoryActionType.LOCATION_REMOVE
})

const removeLocation = (location: Location) =>
    (dispatch: Dispatch<LocationRemoveAction>) => dispatch(locationRemoveAction(location))

// Action mapping ----------------------------------------------------------------------------------

export interface HistoryDispatchProps {
    reloadLocation: typeof reloadLocation
    removeLocation: typeof removeLocation
}

export const mapHistoryDispatchToProps = {
    reloadLocation,
    removeLocation
}
