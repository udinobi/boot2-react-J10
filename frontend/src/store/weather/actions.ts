import { ActionCreator } from "redux"

import {
    WeatherActionType, WeatherData, WeatherDataFailureAction,
    WeatherDataRetrieveAction, WeatherDataRetrievedAction } from "./types"

import { Location } from "../location/city/types"

// WeatherDataFailureAction ------------------------------------------------------------------------

export const weatherDataFailure: ActionCreator<WeatherDataFailureAction> = () => ({
    type: WeatherActionType.WEATHER_DATA_FAILURE
})

// WeatherDataRetrieveAction -----------------------------------------------------------------------

export const weatherDataRetrieve: ActionCreator<WeatherDataRetrieveAction> =
    (location: Location, owmApiKey: string) => ({
        payload: { location, owmApiKey },
        type: WeatherActionType.WEATHER_DATA_RETRIEVE
    })

// WeatherDataRetrievedAction ----------------------------------------------------------------------

export const weatherDataRetrieved: ActionCreator<WeatherDataRetrievedAction> =
    (location: Location, weatherData: WeatherData) => ({
        payload: { location, weatherData },
        type: WeatherActionType.WEATHER_DATA_RETRIEVED
    })

// Action mapping ----------------------------------------------------------------------------------

export interface WeatherDispatchProps {
    weatherDataRetrieve: typeof weatherDataRetrieve
}

export const mapWeatherDispatchToProps = {
    weatherDataRetrieve
}
