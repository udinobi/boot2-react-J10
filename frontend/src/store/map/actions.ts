import { reloadLocation } from '../history/actions'
import { locationSelected } from '../location/cityActions'

export interface MapDispatchProps {
    locationSelected: typeof locationSelected
    reloadLocation: typeof reloadLocation
}

export const mapMapDispatchToProps = {
    locationSelected,
    reloadLocation
}
