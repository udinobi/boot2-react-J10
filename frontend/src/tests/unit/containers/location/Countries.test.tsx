import React from "react"
import * as Renderer from 'react-test-renderer'

import  { Countries } from "../../../../containers/location/Countries"

import { CountriesProps } from "../../../../store/location/country/actions"

describe("--> Snapshot for Containers > location > 'Countries'", () => {
    let props: CountriesProps

    beforeEach(() => {
        props = {
            countries: [],
            countrySelected: jest.fn(),
            loadCountries:  jest.fn()
        }
    })

    it("matches", () => {
        const component = Renderer.create(
            <Countries {...props} />    
        )

        const tree = component.toJSON()
        expect(tree).toMatchSnapshot();
    });
})
