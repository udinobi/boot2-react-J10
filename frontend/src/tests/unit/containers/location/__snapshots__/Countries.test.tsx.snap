// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`--> Snapshot for Containers > location > 'Countries' matches 1`] = `
<div
  className="sc-bdVaJa gikoeI ant-select ant-select-enabled"
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={undefined}
  onFocus={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={undefined}
  onMouseLeave={undefined}
  onTouchStart={[Function]}
  style={undefined}
>
  <div
    aria-autocomplete="list"
    aria-expanded={false}
    aria-haspopup="true"
    className="ant-select-selection
            ant-select-selection--single"
    onKeyDown={[Function]}
    role="combobox"
    tabIndex={0}
  >
    <div
      className="ant-select-selection__rendered"
    >
      <div
        className="ant-select-selection-selected-value"
        style={
          Object {
            "display": "block",
            "opacity": 1,
          }
        }
        title=""
      >
        
      </div>
    </div>
    <span
      className="ant-select-arrow"
      onClick={[Function]}
      style={
        Object {
          "WebkitUserSelect": "none",
          "userSelect": "none",
        }
      }
      unselectable="on"
    >
      <b />
    </span>
  </div>
</div>
`;
